### PR TITLE
Add debug printf validation object

### PR DESCRIPTION
**Pull request by Tony-LunarG**
Processes debugPrintfEXT ops in shaders and returns formatted strings to debug callback or stdout 

docs/debug_printf.md describing use and limitations to follow.

### CODE REVIEW

File: docs/debug_printf.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+[2]: https://www.khronos.org/vulkan/
+
+# Debug Printf
+
+[![Creative Commons][3]][4]
+
+[3]: https://i.creativecommons.org/l/by-nd/4.0/88x31.png "Creative Commons License"
+[4]: https://creativecommons.org/licenses/by-nd/4.0/
+
+Debug Printf is implemented in the SPIR-V Tools optimizer and the `VK_LAYER_KHRONOS_validation` layer.
```
Comment (KarenGhavam-lunarG) [inline docs/debug_printf.md]: Add a sentence describing the benefit of debug printf?

File: docs/debug_printf.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+Debug Printf is an object in the KHRONOS_validation layer, so the VK_LAYER_KHRONOS_validation layer must be loaded.
+See the LAYER_CONFIGURATION document for information on enabling the VK_LAYER_KHRONOS_validation layer.
+Validation itself is not necessary for Debug Printf and can be disabled without affecting Debug Printf functionality.
+
+Debug Printf can be enabled either through a vk_layer_settings.txt file that must be in the program's working directory, or by setting an environment variable �Within a settings file, specify:
+khronos_validation.enables = VK_VALIDATION_FEATURE_ENABLE_DEBUG_PRINTF_EXT or set VK_LAYER_ENABLES=VK_VALIDATION_FEATURE_ENABLE_DEBUG_PRINTF_EXT as an environment variable.
+
+Debug Printf has been implemented as a state tracker validation object, as a peer to gpu assisted validation.
+Because of this, and coupled with the philosophy that validation objects will not communicate with each other, one should never enable both gpu assisted validation and shader printf at the same time.
```
Comment (KarenGhavam-lunarG) [inline docs/debug_printf.md]: gpu assisted validation -> GPU Assisted Validation

Comment (KarenGhavam-lunarG) [inline docs/debug_printf.md]: gpu assisted validation -> GPU Assisted Validation

Comment (Tony-LunarG) [reply docs/debug_printf.md]: Done

File: layers/debug_printf.cpp [hunk 3]
```diff
@@ -0,0 +1,785 @@
...
+            "WARNING - Debug Printf message was truncated, likely due to a buffer size that was too small for the message");
+    }
+    memset(debug_output_buffer, 0, 4 * (debug_output_buffer[0] + 1));
+}
+
+#if defined(__GNUC__)
+#pragma GCC diagnostic pop
+#endif
+
+void DebugPrintf::SendStringToCallback(std::vector<VkLayerDbgFunctionState> debug_callback_list, std::string shader_message) {
```
Comment (mark-lunarg) [inline layers/debug_printf.cpp]: Hey Tony, can we just call LogWarning() instead of duplicating this source?

**Comment by mark-lunarg** [comment]
I'm sure you're already on it, but the last commit will need to get moved up or folded into an earlier commit to keep the intra-commits building correctly.

**Pull request merged**